helix-term/src/ui/picker.rs

```diff
@@ -60,14 +70,45 @@ impl<T> FilePicker<T> {
             })
     }
 
-    fn calculate_preview(&mut self, editor: &Editor) {
-        if let Some((path, _line)) = self.current_file(editor) {
-            if !self.preview_cache.contains_key(&path) && editor.document_by_path(&path).is_none() {
-                // TODO: enable syntax highlighting; blocked by async rendering
-                let doc = Document::open(&path, None, Some(&editor.theme), None).unwrap();
-                self.preview_cache.insert(path, doc);
-            }
+    /// Get preview from preivew cache.
+    /// If not cached, calculate and cache preview, then return reference of it.
+    fn get_preview<'a, 'b>(
+        &'a mut self,
+        path: impl AsRef<std::path::Path>,
+        editor: &'b Editor,
+    ) -> &'a Preview {
+        let path = path.as_ref();
+
+        if self.preview_cache.contains_key(path) && editor.document_by_path(path).is_some() {
+            return self.preview_cache.get(path).unwrap();
         }
+
+        /// Biggest file size to preview in bytes
+        const MAX_PREVIEW_SIZE: u64 = 10 * 1024 * 1024;
+        let data = std::fs::File::open(path).and_then(|file| {
+            let metadata = file.metadata()?;
+            // Read up to 1kb to detect the content type
+            let n = file.take(1024).read_to_end(&mut self.read_buffer)?;
+            let content_type = content_inspector::inspect(&self.read_buffer[..n]);
+            self.read_buffer.clear();
+            Ok((metadata, content_type))
+        });
+        let preview = data
+            .map(
+                |(metadata, content_type)| match (metadata.len(), content_type) {
+                    (_, content_inspector::ContentType::BINARY) => Preview::Binary,
+                    (size, _) if size > MAX_PREVIEW_SIZE => Preview::LargeFile,
+                    _ => {
+                        // TODO: enable syntax highlighting; blocked by async rendering
+                        Document::open(path, None, Some(&editor.theme), None)
+                            .map(Preview::Document)
+                            .unwrap_or(Preview::NotFound)
+                    }
+                },
+            )
+            .unwrap_or(Preview::NotFound);
+        self.preview_cache.insert(path.to_owned(), preview);
+        self.preview_cache.get(path).unwrap()
```

Comment:
This is same as 

```suggestion
        self.preview_cache[path]
```

(it will unwrap internally)